### PR TITLE
Manage access requests programmatically

### PR DESCRIPTION
**Pull request by Wauplin**
Solves https://github.com/huggingface/huggingface_hub/issues/1535.

Let's add official support to programmatically manage access requests for gated repos.

The API defines a `AccessRequest` dataclass with username, email, status, timestamp and fields + 7 endpoints. I reused as much as possible the vocabulary used in the UI. We have:
- `list_pending_access_requests` => returns "pending" list
- `list_accepted_access_requests` => returns "accepted" list
- `list_rejected_access_requests` => returns "rejected" list
- `cancel_access_request` => moves request to "pending" list
- `accept_access_request` => moves request to "accepted" list  
- `reject_access_request` => moves request to "rejected" list
- `grant_access` => add a user to "accepted" list (the user doesn't have to request access first)

Methods docstrings are documented with examples. I don't think it's worth creating a guide dedicated to it in `huggingface_hub` as it would be quite redundant with the Hub docs already. I suggest to update the Hub docs instead to mention the `huggingface_hub` implementation but that's it. 

TODO:
- [ ] update https://huggingface.co/docs/hub/models-gated and https://huggingface.co/docs/hub/datasets-gated once this PR is shipped

### CODE REVIEW

File: src/huggingface_hub/hf_api.py
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:7812]: > and cannot access the community tab

Are you sure about this? I thought all users had access to community tab

**Review by osanseviero — APPROVED** [review]
Nice!

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:7812]: Wow sorry, don't know from where I invented that. I think I'm hallucinating facts :face_with_spiral_eyes:

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:342]: yeah, let's not assume a type. Changed to `Dict[str, Any]`

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: Changed in https://github.com/huggingface/huggingface_hub/pull/1905/commits/31a35dab75e7bfb4b75c26a0d904df887b007780

File: src/huggingface_hub/hf_api.py
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:7945]: 403 if auth error :stuck_out_tongue: (but maybe that's out of scope of what you're aiming for here)

File: tests/test_hf_api.py
Comment (LysandreJik) [inline tests/test_hf_api.py:3692]: Intensive test :ok_hand:

**Review by LysandreJik — APPROVED** [review]
That's awesome! Very cool API.

Is it worth mentioning what kind of permissions they'll need on the repo to have access to the accepted/pending/rejected requests (is it read/write/admin?)? Especially given the example is Llama-2-7b.

**Comment by Wauplin** [comment]
Thanks both for the review!

I have addressed your comments. @LysandreJik regarding roles in organization, the user must have "write" or "admin" role. I have updated the docstrings accordingly. (fyi, role `contributor` also have access to these settings on repos for which they have write access. It's a corner case so I haven't documented it).

I'll wait for the CI to pass and then merge this!

File: src/huggingface_hub/hf_api.py
Comment (julien-c) [reply src/huggingface_hub/hf_api.py]: you used a LLM copilot maybe

Comment (osanseviero) [reply src/huggingface_hub/hf_api.py]: LM stands for Lucain Model

**Comment by RonanKMcGovern** [comment]
@lappemic @Wauplin has something changed here in the last days? After April 18th, my automations querying access seem to have stopped working. Curl requests like this:
```
curl -X GET \             
  'https://huggingface.co/api/repos/Trelis/Mistral-7B-Instruct-v0.1-Summarize-64k/user-access-request/pending' \
  -H 'Authorization: Bearer TOKEN'
```
seem not to be working any more

**Comment by Wauplin** [comment]
> has something changed here in the last days?

@RonanKMcGovern shouldn't be the case no.

If you are getting errors, could you open a new ticket and provide more context (requests/script + error message + stacktrace if applicable). Thanks in advance!

**Comment by RonanKMcGovern** [comment]
Ah very sorry @Wauplin , I just realised I rolled my HF token after a live stream last week without realising it was being used in some automations. Thanks for the quick reply.

**Comment by Wauplin** [comment]
Glad your issue's solved then!

**Pull request merged**
